### PR TITLE
Improve safety of the C interface implementation

### PR DESCRIPTION
**Pull request by bjorn3**

### CODE REVIEW

File: bzlib.rs [hunk 1]
```diff
@@ -591,44 +586,38 @@ unsafe fn BZ2_bzCompressInitHelp(
...
+                arr2.dealloc(bzfree, (*strm).opaque);
         }
 
-        if let Some(ptr) = ftab_alloc {
-            (bzfree)((*strm).opaque, ptr as *mut c_void);
+            if let Some(mut ftab) = ftab {
+                ftab.dealloc(bzfree, (*strm).opaque);
         }
 
-        if !s.is_null() {
```
Comment (bjorn3) [inline bzlib.rs:619]: We are already dereferencing this pointer before this point anyway, so `s` has to be a non-NULL pointer to not have UB. No need to check again.

File: bzlib.rs [hunk 2]
```diff
@@ -1003,21 +992,9 @@ pub unsafe extern "C" fn BZ2_bzCompressEnd(strm: *mut bz_stream) -> c_int {
...
-        (bzfree)(strm.opaque, ptr.cast::<c_void>());
-    }
-    if !(s.ftab).is_null() {
-        let ftab = mem::replace(&mut s.ftab, Ftab::from_ptr(ptr::null_mut()));
-        let ptr = ftab.into_ptr();
-        (bzfree)(strm.opaque, ptr.cast::<c_void>());
-    }
+    s.arr1.dealloc(bzfree, strm.opaque);
+    s.arr2.dealloc(bzfree, strm.opaque);
+    s.ftab.dealloc(bzfree, strm.opaque);
```
Comment (bjorn3) [inline bzlib.rs:1002]: `dealloc` checks if the value exists before deallocating.

File: bzlib.rs [hunk 3]
```diff
@@ -182,7 +182,7 @@ pub(crate) const BZ_N_OVERSHOOT: usize = (BZ_N_RADIX + BZ_N_QSORT + BZ_N_SHELL +
 pub(crate) const FTAB_LEN: usize = u16::MAX as usize + 2;
 
 pub(crate) struct EState {
-    pub strm: *mut bz_stream,
+    pub strm: usize, // Only for a consistency check
```
Comment (folkertdev) [inline bzlib.rs:185]: we could rename this to `strm_addr` then?

**Review by folkertdev — APPROVED** [review]

File: bzlib.rs (hunk 3 above)
Comment (bjorn3) [reply bzlib.rs:185]: Sure

**Pull request merged**